### PR TITLE
feat: carry over promotions toggle on exchanges

### PR DESCRIPTION
**Pull request by fPolic**
## Summary

**What** — What changes are introduced in this PR?

Add ability to apply/remove promotions from outbound exchange item.

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

---

https://github.com/user-attachments/assets/fb473db4-97bf-4bef-81c9-ee4ced6f90e3

---

## TODO

- [x] finish and expand testing
- [x] revamp how inbound/outbound total is calculated on the FE
- [x] cover few more missing edge cases
- [x] fix loading of intial flag state 

---

CLOSES CORE-1291

---

> [!NOTE]
> Adds a toggle to carry over promotions on exchanges/order changes, wiring DB, workflows, API/SDK, admin UI, and tests to apply/remove adjustments accordingly.
> 
> - **Backend/Core**:
>   - **Workflows/Steps**: Introduce `compute-adjustments-for-preview` (generalized), `on-carry-promotions-flag-set`, `update-order-change` workflow, and `list-order-change-actions-by-type` step; compute or delete `ITEM_ADJUSTMENTS_REPLACE` based on `order_change.carry_over_promotions`.
>   - **Integration**: Hook into exchange and order-edit flows to recompute adjustments when items change or flag toggles.
>   - **API**: Add `POST /admin/order-changes/:id` to update `carry_over_promotions` and return `order_change`.
>   - **Types**: Extend `AdminOrderChange` and mutations/responses with `carry_over_promotions`; add `AdminOrderChangeResponse`.
>   - **DB/Model**: Add nullable boolean `carry_over_promotions` to `order_change` (migration + model + query config).
> - **SDK**:
>   - Add `sdk.admin.order.updateOrderChange(id, { carry_over_promotions })`.
> - **Admin Dashboard (UI)**:
>   - Add "Carry over promotions" switch in exchange create form with tooltip; default state from `preview.order_change` and updates via new mutation; adjust totals calculation; show applied promo codes icon in order summary.
>   - Add `useUpdateOrderChange` hook; i18n strings/schema updates.
> - **Tests**:
>   - Integration tests for exchanges toggling promotions (enabled/disabled before request) and order edits enabling flag; verify preview/final adjustments and totals.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 44398cd0b4347b74b3f277d77a5865d0a716d1b6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+      for (const promotion of order.promotions) {
+        const applicationMethod = (promotion as any).application_method
+
+        if (!applicationMethod) {
+          continue
+        }
+
+        const allocation = applicationMethod.allocation
+        const type = applicationMethod.type
+
```
Comment (fPolic) [inline packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:71]: self-review(todo): if allocation is other than `EACH` | `ACROSS` (e.g. `ONCE`) throw early

File: packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/exchange-create-form.tsx [hunk 2]
```diff
@@ -498,6 +522,50 @@ export const ExchangeCreateForm = ({
...
+                              }}
+                              {...field}
+                            />
+                          </Form.Control>
+                          <div className="block">
+                            <Form.Label>
+                              {t("orders.exchanges.carryOverPromotion")}
+                            </Form.Label>
+                            <Form.Hint className="!mt-1">
+                              {t("orders.exchanges.carryOverPromotionHint")}
```
Comment (olivermrbl) [inline packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/exchange-create-form.tsx:567]: nit: can we add a tooltip that also explains what types of promotions this is applicable to?

File: packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts [hunk 3]
```diff
@@ -19,6 +19,7 @@ export const ExchangeCreateSchema = z.object({
   inbound_option_id: z.string().nullish(),
   outbound_option_id: z.string().nullish(),
   send_notification: z.boolean().optional(),
+  carry_over_promotion: z.boolean().optional(),
```
Comment (olivermrbl) [inline packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts]: nit: believe this should be pluralized for consistency no? `_promotions`

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+          allocation !== ApplicationMethodAllocation.ACROSS &&
+          allocation !== ApplicationMethodAllocation.EACH
+        ) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Promotion ${
+              promotion.code || promotion.id
+            } has invalid allocation. Only promotions with EACH or ACROSS allocation can be carried over to outbound exchange items.`
+          )
+        }
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:82]: thought: I wonder if we should skip the promotions instead of throwing? If you have multiple promotions applied, you cannot carry over some of them.

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      }
+
+      if (invalidPromotions.length > 0) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Promotions with codes ${invalidPromotions.join(
+            ", "
+          )} have invalid allocation. Fixed promotions must have EACH allocation, and percentage promotions must have EACH or ACROSS allocation.`
+        )
+      }
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:109]: Same as above

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    const updatedOrderChange = transform(
+      { updatedOrderChanges },
+      ({ updatedOrderChanges }) => updatedOrderChanges[0]
+    )
+
+    when(
+      "should-apply-promotions",
+      { flag: input.carry_over_promotions },
+      ({ flag }) => flag === true
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts]:
```suggestion
      ({ flag }) => !!flag
```

File: packages/medusa/src/api/admin/order-changes/[id]/route.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import { AdminPostOrderChangesReqSchemaType } from "../validators"
+
+export const POST = async (
+  req: AuthenticatedMedusaRequest<AdminPostOrderChangesReqSchemaType>,
+  res: MedusaResponse<HttpTypes.AdminOrderChangeResponse>
+) => {
+  const { id } = req.params
+  const { carry_over_promotions } = req.validatedBody
+
+  const workflow = setCarryOverPromotionFlagForOrderChangeWorkflow(req.scope)
```
Comment (olivermrbl) [inline packages/medusa/src/api/admin/order-changes/[id]/route.ts]: nit: feels a bit off to use a very workflow in a generic update API route, should we create (or rename) an `updateOrderChangesWorkflow` instead?

File: packages/medusa/src/api/admin/order-changes/[id]/route.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  await workflow.run({
+    input: {
+      order_change_id: id,
+      carry_over_promotions,
+    },
+  })
+
+  const orderChange = await refetchEntity({
```
Comment (olivermrbl) [inline packages/medusa/src/api/admin/order-changes/[id]/route.ts]: nit: use query graph for consistency

File: packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts (hunk 3 above)
Comment (olivermrbl) [reply packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts]: wonder how this has worked actually? 😅

**Review by olivermrbl — COMMENTED** [review]
Overall LGTM, few nits

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts (hunk 4 above)
Comment (fPolic) [reply packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:82]: yup sounds reasonable, should we combine this with the UI tooltip and state which promotions can be carried over?

File: packages/medusa/src/api/admin/order-changes/[id]/route.ts (hunk 7 above)
Comment (fPolic) [reply packages/medusa/src/api/admin/order-changes/[id]/route.ts]: yeah this was initially proposed in the RFC, but wanted to leave that when there is a need for a generic update - will refactor!

File: packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts (hunk 3 above)
Comment (fPolic) [reply packages/admin/dashboard/src/routes/orders/order-create-exchange/components/exchange-create-form/schema.ts]: good catch, I refactored the name after initial implementation and forgot to change it here

File: packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts (hunk 4 above)
Comment (olivermrbl) [reply packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:82]: Hmm, I think for now, I would be fine just having the tooltip. I fear it can get a bit cluttered

Comment (fPolic) [reply packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:82]: @olivermrbl on second thought, this would affect other items on the order. If an order has 2 promotions, one with `ONCE` and one with `EACH` allocation, after exchange, only the `EACH` one would be effectively applied after exchange confirmation? `ONCE` adjustments would be lost since we cannot selectively apply a subset of promotions to one subset of items and maintain that across edits.

Comment (olivermrbl) [reply packages/core/core-flows/src/order/workflows/set-carry-over-promotion-flag.ts:82]: That does make sense yeah. It is probably an edge case to have more than 1 promotion anyway, so think it's fine to keep it as is and thro.

**Review by olivermrbl — APPROVED** [review]
Nice LGTM!

**Comment by fPolic** [comment]
tought: now when exchange is confirmed, and promotions are not carried over, it's not clear from the summary on the order details how the promotion has been applied. We might consider adding an indicator with promotions per line item as we did for edit items recently, wdyt - something to tackle in a follow up?

**Review by olivermrbl — APPROVED** [review]
Feel free to merge
